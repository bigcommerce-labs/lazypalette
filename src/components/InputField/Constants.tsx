export const enum Status {
    Invalid,
    Valid,
    Undefined,
}

export const enum Notes {
    Required = 'Required*',
}
