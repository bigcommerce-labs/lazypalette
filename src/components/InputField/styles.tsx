import { theme } from 'pattern-lab';
import { Status } from './Constants';
import styled, { StyledComponentClass } from 'styled-components';
import { colors } from '../../styleConstants';

const { colors: patternLabColors } = theme;
const defaultBorderColor = colors.stroke;
const defaultBorderColorOnHover = colors.darkBlueMagentaishGray;
const immutableBorderColorOnFocus = 'none';
const immutableBorderColorOnHover = 'none';
const immutableInputBackgroundColor = colors.cyanishWhite;
const mutableInputBackgroundColor = colors.white;
const statuslessImmutableBorderColor = colors.cyanBluishWhite;
const warningBorderColor = colors.lightBrilliantOrangeYellow;

export const Container = styled.div`
    display: flex;
    flex-direction: column;
`;

export const Label = styled.label`
    padding: 5px 1px;
    color: colors.primary;
`;

interface SmallProps {
    required: boolean;
}

export const Small = styled.small`
    color: ${(props: SmallProps) => props.required ? colors.lightBrilliantRedOrange : 'initial'};
    font-size: 12px;
    padding-top: 4px;
`;

const getBackgroundColor = (props: InputFieldProps) => {
    if (props.readonly || props.disabled) {
        return immutableInputBackgroundColor;
    } else {
        return mutableInputBackgroundColor;
    }
};

const getBorderColor = (props: InputFieldProps) => {
    if (props.status !== Status.Undefined) {
        switch (props.status) {
            case Status.Invalid: return patternLabColors.status.error;
            case Status.Valid: return patternLabColors.status.success;
            default: return defaultBorderColor;
        }
    } else if (props.warning) {
        return warningBorderColor;
    } else if (props.readonly || props.disabled) {
        return statuslessImmutableBorderColor;
    } else {
        return defaultBorderColor;
    }
};

const getBorderColorOnHover = (props: InputFieldProps) => {
    if (props.readonly || props.disabled) {
        return immutableBorderColorOnHover;
    } else {
        return defaultBorderColorOnHover;
    }
};

const getBorderColorOnFocus = (props: InputFieldProps) => {
    if (props.readonly || props.disabled) {
        return immutableBorderColorOnFocus;
    } else {
        return patternLabColors.brand.primary;
    }
};

interface InputFieldProps {
    disabled?: boolean;
    readonly?: boolean;
    status?: Status;
    warning?: boolean;
}

export const InputField: StyledComponentClass<any, any> = styled.input`
    cursor: ${(props: InputFieldProps) => props.disabled ? 'not-allowed' : 'text'};
    background-color: ${(props: InputFieldProps) => getBackgroundColor(props)};
    border: 1px solid ${(props: InputFieldProps) => getBorderColor(props)};
    border-radius: 2px;
    box-sizing: border-box;
    height: 35px;
    padding: 10px;
    width: 100%;

    :hover {
      border-color: ${(props: InputFieldProps) => getBorderColorOnHover(props)}
    }

    :focus {
      border-color: ${(props: InputFieldProps) => getBorderColorOnFocus(props)}
      outline: none;
    }
  `;
