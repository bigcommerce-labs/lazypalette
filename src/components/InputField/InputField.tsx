import React, {
    ChangeEvent,
    ChangeEventHandler,
    Component,
    FocusEvent,
    FocusEventHandler,
    SFC,
} from 'react';
import uuid from 'uuid';

import { Container, InputField as StyledInputField, Label, Small } from './styles';
import { Notes, Status } from './Constants';

interface LabelProps {
    children?: string;
    htmlFor?: string;
    note?: string;
    required?: boolean;
}

const InputFieldLabel: SFC<LabelProps> = props => {
    const { children: label, htmlFor } = props;

    return (
        <Label htmlFor={htmlFor}>
            {label}
        </Label>
    );
};

export interface InputFieldState {
    status: Status;
    value: string | undefined;
}

export interface InputFieldProps {
    defaultValue?: any;
    disabled?: boolean;
    inputId?: string;
    inset?: string;
    label?: string;
    note?: string;
    onBlur?: FocusEventHandler<HTMLInputElement>;
    onChange?: ChangeEventHandler<HTMLInputElement>;
    pattern?: string | RegExp;
    placeholder?: string;
    postfix?: string;
    prefix?: string;
    readonly?: boolean;
    required?: boolean;
    type?: string | undefined;
    warning?: boolean;
}

export default class InputField extends Component<InputFieldProps, InputFieldState> {
    static defaultProps = {
        disabled: false,
        inputId: uuid(),
        onBlur: () => undefined,
        onChange: () => undefined,
        readonly: false,
        required: false,
        warning: false,
    };

    readonly state: InputFieldState = {
        status: Status.Undefined,
        value: '',
    };

    constructor(props: InputFieldProps) {
        super(props);
        this.state = {
            status: Status.Undefined,
            value: props.defaultValue,
        };
    }

    componentWillMount() {
        if (this.props.pattern) {
            this.validate();
        }
    }

    render() {
        const { status } = this.state;
        const {
            defaultValue,
            disabled,
            inputId,
            label,
            note = this.prepareNote(),
            pattern,
            placeholder,
            readonly,
            required,
            type,
            warning,
        } = this.props;

        return (
            <Container>
                <InputFieldLabel htmlFor={inputId}>
                    {label}
                </InputFieldLabel>
                <StyledInputField
                    defaultValue={defaultValue}
                    disabled={disabled!}
                    id={inputId}
                    onChange={this.onChange}
                    onBlur={this.onBlur}
                    pattern={pattern!}
                    placeholder={placeholder!}
                    readonly={readonly!}
                    required={required!}
                    status={status}
                    type={type}
                    warning={warning!}
                />
                {note &&
                    <Small required={required!}>
                        {note}
                    </Small>
                }
            </Container>
        );
    }

    private onBlur = (event: FocusEvent<HTMLInputElement>) => {
        this.props.onBlur!(event);

        if (this.props.required && !this.state.value) {
            return this.setState({ status: Status.Invalid });
        }

        if (this.props.pattern) {
            this.validate();
        }
    };

    private onChange = (event: ChangeEvent<HTMLInputElement>) => {
        this.setState({
            value: event.target.value,
        });

        this.props.onChange!(event);
    };

    private prepareNote() {
        if (this.props.note) {
            return this.props.note;
        } else if (this.props.required && !this.state.value) {
            return Notes.Required;
        } else {
            return undefined;
        }
    }

    private validate = () => {
        this.setState((prevState: Readonly<InputFieldState>, props: InputFieldProps) => {
            const regExp = props.pattern instanceof RegExp ? props.pattern : new RegExp(props.pattern!, 'u');

            return {
                status: regExp.test(prevState.value!) ? Status.Valid : Status.Invalid,
            };
        });
    };
}
