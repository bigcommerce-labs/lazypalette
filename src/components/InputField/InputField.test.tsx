import { shallow } from 'enzyme';
import React from 'react';

import InputField from './InputField';

const id = 'InputTestId';

it('renders', () => {
  const inputField = shallow(
    <InputField inputId={id} label="Name your version" />
  );
  expect(inputField).toMatchSnapshot();
});

it('renders a valid state', () => {
  const inputField = shallow(
    <InputField inputId={id} label="Name your version" defaultValue="abc" pattern="abc" />
  );
  expect(inputField).toMatchSnapshot();
});

it('renders an invalid state', () => {
  const inputField = shallow(
    <InputField inputId={id} label="Name your version" defaultValue="adlkfj" pattern="abc" />
  );
  expect(inputField).toMatchSnapshot();
});
