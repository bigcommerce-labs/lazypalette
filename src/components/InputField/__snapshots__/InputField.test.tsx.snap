// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<styled.div>
  <InputFieldLabel
    htmlFor="InputTestId"
  >
    Name your version
  </InputFieldLabel>
  <styled.input
    disabled={false}
    id="InputTestId"
    onBlur={[Function]}
    onChange={[Function]}
    readonly={false}
    required={false}
    status={2}
    warning={false}
  />
</styled.div>
`;

exports[`renders a valid state 1`] = `
<styled.div>
  <InputFieldLabel
    htmlFor="InputTestId"
  >
    Name your version
  </InputFieldLabel>
  <styled.input
    defaultValue="abc"
    disabled={false}
    id="InputTestId"
    onBlur={[Function]}
    onChange={[Function]}
    pattern="abc"
    readonly={false}
    required={false}
    status={1}
    warning={false}
  />
</styled.div>
`;

exports[`renders an invalid state 1`] = `
<styled.div>
  <InputFieldLabel
    htmlFor="InputTestId"
  >
    Name your version
  </InputFieldLabel>
  <styled.input
    defaultValue="adlkfj"
    disabled={false}
    id="InputTestId"
    onBlur={[Function]}
    onChange={[Function]}
    pattern="abc"
    readonly={false}
    required={false}
    status={0}
    warning={false}
  />
</styled.div>
`;
